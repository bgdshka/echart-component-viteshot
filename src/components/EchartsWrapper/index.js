import React, { useRef, useEffect } from "react";
import { init, getInstanceByDom } from "echarts";

/**
 * Wrapper of `echarts` instance for react
 * @param {Object} $.option - chart data 
 * @param {Object?} $.style - for inline style
 * @param {Object?} $.settings - additional parameters to merge
 * @param {Boolean?} $.loading - to display loading status
 * @param {Object?} $.theme - for `echarts` themes
 */
function EchartsWrapper({ option, style, settings, loading, theme }) {
  const chartRef = useRef(null);

  useEffect(() => {
    // Initialize chart
    let chart;
    if (chartRef.current !== null) {
      chart = init(chartRef.current, theme);
    }

    // Add chart resize listener
    // ResizeObserver is leading to a bit janky UX
    function resizeChart() {
      chart?.resize();
    }
    window.addEventListener("resize", resizeChart);

    // Return cleanup function
    return () => {
      chart?.dispose();
      window.removeEventListener("resize", resizeChart);
    };
  }, [theme]);

  useEffect(() => {
    // Update chart
    if (chartRef.current !== null) {
      const chart = getInstanceByDom(chartRef.current);
      chart.setOption(option, settings);
    }
  }, [option, settings, theme]); // Whenever theme changes we need to add option and setting due to it being deleted in cleanup function

  useEffect(() => {
    // Update chart
    if (chartRef.current !== null) {
      const chart = getInstanceByDom(chartRef.current);
      loading === true ? chart.showLoading() : chart.hideLoading();
    }
  }, [loading, theme]);

  return <div ref={chartRef} style={{ width: "100%", height: "100px", ...style }} />;
}

export default EchartsWrapper;
