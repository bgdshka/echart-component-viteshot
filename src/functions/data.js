// There may be different functions here
// or a more complex function to convert data to options for echarts.
// Depends on the task/subject domain.

/**
 * Convert series data to options in week interval for `echarts`.
 * @param {Array[Object]} series - array series
 * @returns {Array} - generated array
 */
export const convertSeriesToOption = (series = []) => {
  return {
    xAxis: {
      type: "category",
      data: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
    },
    yAxis: {
      type: "value",
    },
    series,
  };
};
