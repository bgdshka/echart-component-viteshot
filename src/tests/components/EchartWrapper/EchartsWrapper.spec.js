import React from "react";
import renderer from "react-test-renderer";
import EchartsWrapper from "../../../components/EchartsWrapper";
import { optionBarMock } from "../../mock/data";

it("Bar chart renders correctly", () => {
  const tree = renderer
    .create(
      <EchartsWrapper
        option={optionBarMock}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
