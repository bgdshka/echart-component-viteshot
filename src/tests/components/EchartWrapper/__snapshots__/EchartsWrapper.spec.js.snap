// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar chart renders correctly 1`] = `
<div
  style={
    Object {
      "height": "100px",
      "width": "100%",
    }
  }
/>
`;
